File: server/src/models/Product.js
import mongoose from "mongoose";

const productSchema = new mongoose.Schema({
  name: {
    type: String,
    required: true,
    trim: true
  },
  category: {
    type: String,
    required: true,
    enum: ["timber", "furniture", "construction"],
    trim: true
  },
  subcategory: {
    type: String,
    trim: true
  },
  quantity: {
    type: Number,
    default: 0,
    min: 0
  },
  unit: {
    type: String,
    enum: ["cubic ft", "pieces", "piece"],
    default: "pieces"
  },
  price: {
    type: Number,
    required: true,
    min: 0
  },
  size: {
    type: String,
    trim: true
  },
  description: {
    type: String,
    trim: true,
    maxlength: 1000
  },
  images: [{
    // Legacy: base64 data URL (when not using Cloudinary)
    data: { type: String },
    // Cloudinary: stored URL and public_id for delete
    url: { type: String },
    publicId: { type: String },
    contentType: {
      type: String,
      required: true
    },
    filename: {
      type: String,
      required: true
    },
    color: {
      type: String,
      trim: true
    },
    // Cover image shown on listing; first image is cover if none set
    isCover: { type: Boolean, default: false }
  }],
  attributes: {
    type: mongoose.Schema.Types.Mixed,
    default: {}
  },
  productType: {
    type: String,
    enum: ["ready-stock", "made-to-order"],
    default: "ready-stock"
  },
  customizationOptions: {
    woodTypes: [{ type: String, trim: true }],
    estimatedProductionTime: { type: String, trim: true },
    basePrice: { type: Number, min: 0 }
  },
  isActive: {
    type: Boolean,
    default: true
  },
  featuredType: {
    type: String,
    enum: ["best", "new", "discount", "none", "featured"],
    default: "none"
  },
  rating: {
    type: Number,
    default: 0,
    min: 0,
    max: 5
  },
  reviewCount: {
    type: Number,
    default: 0,
    min: 0
  },
  warrantyIncluded: {
    type: Boolean,
    default: false
  },
  warrantyMonths: {
    type: Number,
    default: 0,
    min: 0
  }
}, {
  timestamps: true
});

// Validate max 5 images and each image has either data (legacy) or url (Cloudinary)
productSchema.pre('save', function (next) {
  if (this.images && this.images.length > 50) {
    return next(new Error('Maximum 50 images allowed per product across all color variants'));
  }
  if (this.images && this.images.length > 0) {
    const invalid = this.images.some(img => !img.data && !img.url);
    if (invalid) {
      return next(new Error('Each image must have either data or url'));
    }
  }
  next();
});

// Index for better query performance
productSchema.index({ category: 1, isActive: 1 });
productSchema.index({ category: 1, subcategory: 1, isActive: 1 });
productSchema.index({ name: 1 });
productSchema.index({ price: 1 });
productSchema.index({ featuredType: 1, isActive: 1 });

const Product = mongoose.model('Product', productSchema);

export default Product;

File: jc_timber_mobile/lib/models/product.dart
import 'dart:convert';

class Product {
  final String id;
  final String name;
  final String category;
  final String? subcategory;
  final int quantity;
  final String unit;
  final double price;
  final String? size;
  final String? description;
  final List<ProductImage> images;
  final Map<String, dynamic> attributes;
  final double rating;
  final int reviewCount;
  final String? featuredType;
  final String? material;
  final String? color;
  final String? brand;
  final String? weight;

  const Product({
    required this.id,
    required this.name,
    required this.category,
    this.subcategory,
    required this.quantity,
    required this.unit,
    required this.price,
    this.size,
    this.description,
    required this.images,
    required this.attributes,
    this.rating = 0,
    this.reviewCount = 0,
    this.featuredType,
    this.material,
    this.color,
    this.brand,
    this.weight,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: (json['_id'] ?? json['id']).toString(),
      name: json['name'] as String? ?? '',
      category: json['category'] as String? ?? '',
      subcategory: json['subcategory'] as String?,
      quantity: (json['quantity'] as num?)?.toInt() ?? 0,
      unit: json['unit'] as String? ?? 'pieces',
      price: (json['price'] as num?)?.toDouble() ?? 0,
      size: json['size'] as String?,
      description: json['description'] as String?,
      images: (json['images'] as List<dynamic>? ?? [])
          .map((e) => ProductImage.fromJson(e as Map<String, dynamic>))
          .toList(),
      attributes:
          (json['attributes'] as Map<String, dynamic>? ?? <String, dynamic>{}),
      rating: (json['rating'] as num?)?.toDouble() ?? 0,
      reviewCount: (json['reviewCount'] as num?)?.toInt() ?? 0,
      featuredType: json['featuredType'] as String?,
      material: json['material'] as String?,
      color: json['color'] as String?,
      brand: json['brand'] as String?,
      weight: json['weight'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'category': category,
      'subcategory': subcategory,
      'quantity': quantity,
      'unit': unit,
      'price': price,
      'size': size,
      'description': description,
      'images': images.map((e) => e.toJson()).toList(),
      'attributes': attributes,
      'rating': rating,
      'reviewCount': reviewCount,
      'featuredType': featuredType,
      'material': material,
      'color': color,
      'brand': brand,
      'weight': weight,
    };
  }

  String? get primaryImageUrl {
    if (images.isEmpty) return null;
    final img = images.first;
    if (img.url != null && img.url!.isNotEmpty) return img.url;
    if (img.data != null && img.data!.isNotEmpty) {
      // Already a full data URL?
      if (img.data!.startsWith('data:')) return img.data;
      // Fallback: assume base64 jpeg
      return 'data:image/jpeg;base64,${img.data}';
    }
    return null;
  }

  String? getImageUrl(ProductImage image) {
    if (image.url != null && image.url!.isNotEmpty) return image.url;
    if (image.data != null && image.data!.isNotEmpty) {
      if (image.data!.startsWith('data:')) return image.data;
      if (image.data!.startsWith('http')) return image.data;
      return 'data:${image.contentType};base64,${image.data}';
    }
    return null;
  }
}

class ProductImage {
  final String? data;
  final String? url;
  final String contentType;
  final String filename;

  const ProductImage({
    this.data,
    this.url,
    required this.contentType,
    required this.filename,
  });

  factory ProductImage.fromJson(Map<String, dynamic> json) {
    return ProductImage(
      data: json['data'] as String?,
      url: json['url'] as String?,
      contentType: json['contentType'] as String? ?? 'image/jpeg',
      filename: json['filename'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'data': data,
      'url': url,
      'contentType': contentType,
      'filename': filename,
    };
  }
}

